### PR TITLE
Add new restroom activity apis

### PR DESCRIPTION
**Pull request by Koro1229**
Add a floating button in MainActivity for update restroom location
Call registerRestroom or updateRestroomLocation when click submit button in AddRestroomActivity
Load latest restrooms' data from backend after registerRestroom or updateRestroomLocation

### CODE REVIEW

File: app/src/main/java/com/example/android/service/BackendClient.java [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
 import okhttp3.Request;
+import okhttp3.RequestBody;
 import okhttp3.Response;
 
 public class BackendClient {
     private static BackendClient instance = null;
     private final OkHttpClient _client;
-    private final String _baseUrl = "https://fa43-219-85-87-19.ap.ngrok.io";
+    private final String _baseUrl = "https://f942-220-135-128-111.jp.ngrok.io";
+    public static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/service/BackendClient.java]: Suggest using `private` in this case.

File: app/src/main/java/com/example/android/utils/BasicAsyncTask.java [hunk 2]
```diff
@@ -20,6 +22,8 @@ protected Void doInBackground(Void... voids) {
     @Override
     protected void onPostExecute(Void unused) {
         super.onPostExecute(unused);
-        _postExecute.run();
+        if(_postExecute != null){
+            _postExecute.run();
+        }
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/utils/BasicAsyncTask.java]: Please format this code.

File: app/src/main/java/com/example/android/view/AddRestroomActivity.java [hunk 3]
```diff
@@ -2,33 +2,61 @@
...
-        restroomIdEditText = findViewById(R.id.restroomIdEditText);
+
+        Intent intent = getIntent();
+        String restroomIntentFlag = intent.getStringExtra(RESTROOM_INTENT_FLAG);
+        if(restroomIntentFlag.equals("register")){
+            setContentView(R.layout.add_restroom_activity);
+        }else if(restroomIntentFlag.equals("update")){
+            setContentView(R.layout.update_restroom_activity);
+            restroomIdEditText = findViewById(R.id.restroomIdEditText);
+        }
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/AddRestroomActivity.java]: Please format this code.

File: app/src/main/java/com/example/android/view/MainActivity.java [hunk 4]
```diff
@@ -16,34 +17,43 @@
...
+        _adapter = new RestroomListAdapter(this);
+        _viewModel = new RestroomsViewModel(_adapter);
 
         _restroomRecyclerView = findViewById(R.id.recyclerView);
-        _restroomRecyclerView.setAdapter(adapter);
+        _restroomRecyclerView.setAdapter(_adapter);
         _restroomRecyclerView.setLayoutManager(new LinearLayoutManager(this));
 
-        new BasicAsyncTask(() -> _viewModel.loadRestroomsData(), () -> adapter.notifyDataSetChanged()).execute();
+        new BasicAsyncTask(() -> _viewModel.loadRestroomsData(), _adapter::notifyDataSetChanged).execute();
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/MainActivity.java]: Could you please help extract this `Runnable`?

```suggestion
        final Runnable backgroundTask = () -> _viewModel.loadRestroomsData();
        new BasicAsyncTask(backgroundTask, _adapter::notifyDataSetChanged).execute();
```

File: app/src/main/java/com/example/android/view/MainActivity.java [hunk 5]
```diff
@@ -16,34 +17,43 @@
...
     }
 
     @Override
     public void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
         if (requestCode == 1) {
             if (resultCode == RESULT_OK) {
-                _restroomRecyclerView.getAdapter().notifyDataSetChanged();
+                new BasicAsyncTask(() -> _viewModel.loadRestroomsData(), _adapter::notifyDataSetChanged).execute();
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/MainActivity.java]: Ditto.

**Review by ricky9667 — COMMENTED** [review]
It seems like you don't have the habit of formatting code after you make changes. You can type `Ctrl + Alt + L` shortcut key to do the job.
I suggest you to always format your code before you commit, however if you forget (like this time) you can add a `style: ` commit to refactor all of them.

**Review by xiao0818 — APPROVED** [review]

File: app/src/main/java/com/example/android/view/AddRestroomActivity.java [hunk 6]
```diff
@@ -2,33 +2,61 @@
 
 import android.content.Intent;
 import android.os.Bundle;
+import android.util.Log;
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/AddRestroomActivity.java]: Suggest removing unused imports.

File: app/src/main/java/com/example/android/view/AddRestroomActivity.java [hunk 7]
```diff
@@ -2,33 +2,61 @@
...
 
 import com.example.android.R;
-import com.example.android.model.Restroom;
 import com.example.android.service.Store;
+import com.example.android.utils.BasicAsyncTask;
+import com.example.android.viewModel.RestroomInfoViewModel;
 
 public class AddRestroomActivity extends AppCompatActivity {
+    private RestroomInfoViewModel _viewModel;
     private final Store store = Store.getInstance();
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/AddRestroomActivity.java]: It seem like `store` is not used in this file anymore, suggest removing it.

File: app/src/main/java/com/example/android/view/AddRestroomActivity.java [hunk 8]
```diff
@@ -2,33 +2,61 @@
...
+
+        setResult(RESULT_OK, replyIntent);
+        finish();
+    }
+
+    public void updateRestroomLocation(View view) {
+        Intent replyIntent = new Intent();
+        final String restroomIdText = restroomIdEditText.getText().toString();
+        final String restroomLocation = restroomLocationEditText.getText().toString();
+        int restroomId = Integer.parseInt(restroomIdText);
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/AddRestroomActivity.java]: Suggest adding `final` to this variable.

File: app/src/main/java/com/example/android/view/MainActivity.java [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 import android.content.Intent;
 import android.os.Bundle;
+import android.util.Log;
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/MainActivity.java]: Ditto.

File: app/src/main/java/com/example/android/view/ToiletActivity.java [hunk 10]
```diff
@@ -16,6 +16,8 @@
 import com.example.android.viewModel.ToiletsViewModel;
 
 public class ToiletActivity extends AppCompatActivity {
+    private final String RESTROOM_ID_EXTRA = "restroomIdExtra";
+    private final String RESTROOM_LOCATION_EXTRA = "restroomLocationExtra";
     private final Store _store = Store.getInstance();  // TODO: remove after all branches have been merge
```
Comment (ricky9667) [inline app/src/main/java/com/example/android/view/ToiletActivity.java]: Ditto.

File: app/src/main/java/com/example/android/viewModel/RestroomInfoViewModel.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            jsonObject.put("restroomId", id);
+            jsonObject.put("location", location);
+
+            String response = _client.updateRestroomLocation(jsonObject.toString());
+            Log.d("updateRestroomLocation", response);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+}
```
Comment (xiao0818) [inline app/src/main/java/com/example/android/viewModel/RestroomInfoViewModel.java]: adding a blank line at the end of this file

**Review by xiao0818 — APPROVED** [review]

File: app/src/main/java/com/example/android/viewModel/RestroomInfoViewModel.java [hunk 12]
```diff
@@ -34,4 +34,5 @@ public void updateRestroomLocation(int id, String location) {
             e.printStackTrace();
         }
     }
-}
\ No newline at end of file
+
```
Comment (xiao0818) [inline app/src/main/java/com/example/android/viewModel/RestroomInfoViewModel.java]: please remove this extra blank line

**Review by ricky9667 — APPROVED** [review]
LGTM.

**Pull request merged**